src/compiler.ts

```diff
@@ -6571,14 +6654,45 @@ export class Compiler extends DiagnosticEmitter {
         return module.unreachable();
       }
     }
-    return this.compileCallIndirect(
-      assert(signature), // FIXME: bootstrap can't see this yet
-      indexArg,
-      expression.arguments,
-      expression,
-      0,
-      contextualType == Type.void
-    );
+    //Once we get here, we have a function reference. With the new scheme, this function
+    //could possibly be a closure. So here we check to see if it's a closure, then apply
+    //the appropriate call logic
+    var returnType = signature.returnType;
+    var tempFunctionReferenceLocal = this.currentFlow.getTempLocal(Type.i32);
+    return module.block(null, [
+      module.local_set(tempFunctionReferenceLocal.index, indexArg),
+      this.ifClosure(
+        module.local_get(tempFunctionReferenceLocal.index, NativeType.I32),
+        this.compileCallIndirect( //If this is a closure
+          assert(signature).toClosureSignature(), // FIXME: asc can't see this yet
+          module.block(null, [
+            module.load(
+              4,
+              true,
+              this.getClosurePtr(
+                module.local_get(tempFunctionReferenceLocal.index, NativeType.I32),
+              ),
+              NativeType.I32,
+              0
+            ),
+          ], this.options.nativeSizeType),
+          expression.arguments,
+          expression,
+          this.getClosurePtr(
+            module.local_get(tempFunctionReferenceLocal.index, NativeType.I32),
+          ),
+          contextualType == Type.void
+        ),
+        this.compileCallIndirect( //If this function isn't a closure
+          assert(signature), // FIXME: asc can't see this yet
+          module.local_get(tempFunctionReferenceLocal.index, NativeType.I32),
+          expression.arguments,
+          expression,
+          0,
+          contextualType == Type.void
+        )
+      )
+    ], constraints & Constraints.WILL_DROP ? contextualType.toNativeType() : returnType.toNativeType())
```

Comment:
Can you explain the relation of the block type to `WILL_DROP` here? Typically, `WILL_DROP` is used in something like `i++` where the compiler can then compile without teeing the value, making the return type `Type.void`, and I would have expected a `void` around here as well.

Reply:
When the result of the call was immediately discarded, then the contextualType would be void, so I take that as the native type. Could probably just replace `contextualType.toNativeType` just with `Type.void`. Otherwise I just used the call return type.